topology/probes/blockdev/blockdev.go

```diff
@@ -0,0 +1,515 @@
+		DiscMax:      blockdev.DiscMax,
+		DiscZero:     p.getBool(blockdev.DiscZero),
+		Fsavail:      blockdev.Fsavail,
+		Fssize:       blockdev.Fssize,
+		Fstype:       blockdev.Fstype,
+		FsusePercent: blockdev.FsusePercent,
+		Fsused:       blockdev.Fsused,
+		Group:        blockdev.Group,
+		Hctl:         blockdev.Hctl,
+		Hotplug:      p.getBool(blockdev.Hotplug),
+		Kname:        blockdev.Kname,
+		Label:        blockdev.Label,
+		LogSec:       p.getInt(blockdev.LogSec),
+		MajMin:       blockdev.MajMin,
+		MinIo:        p.getInt(blockdev.MinIo),
+		Mode:         blockdev.Mode,
+		Model:        blockdev.Model,
+		Mountpoint:   blockdev.Mountpoint,
+		OptIo:        p.getInt(blockdev.OptIo),
+		Owner:        blockdev.Owner,
+		Partflags:    blockdev.Partflags,
+		Partlabel:    blockdev.Partlabel,
+		Parttype:     blockdev.Parttype,
+		Partuuid:     blockdev.Partuuid,
+		Path:         blockdev.Path,
+		PhySec:       p.getInt(blockdev.PhySec),
+		Pttype:       blockdev.Pttype,
+		Ptuuid:       blockdev.Ptuuid,
+		Ra:           p.getInt(blockdev.Ra),
+		Rand:         p.getBool(blockdev.Rand),
+		Rev:          blockdev.Rev,
+		Rm:           p.getBool(blockdev.Rm),
+		Ro:           p.getBool(blockdev.Ro),
+		Rota:         p.getBool(blockdev.Rota),
+		RqSize:       p.getInt(blockdev.RqSize),
+		Sched:        blockdev.Sched,
+		Serial:       blockdev.Serial,
+		Size:         blockdev.Size,
+		State:        blockdev.State,
+		Subsystems:   blockdev.Subsystems,
+		Tran:         blockdev.Tran,
+		Type:         blockdev.Type,
+		UUID:         blockdev.UUID,
+		Vendor:       blockdev.Vendor,
+		Wsame:        blockdev.Wsame,
+		WWN:          WWN,
+	}
+	metadata = graph.Metadata{
+		"Path":     p.getPath(blockdev),
+		"Type":     nodeType,
+		"Name":     p.getName(blockdev),
+		"Manager":  managerType,
+		"BlockDev": blockdevMetadata,
+	}
+
+	return metadata
+}
+
+func (p *ProbeHandler) findBlockDev(path string, id string, newDevInfo []BlockDevice) bool {
+	for i := range newDevInfo {
+		if id == p.getID(newDevInfo[i]) && path == p.getPath(newDevInfo[i]) {
+			return true
+		}
+		if p.findBlockDev(path, id, newDevInfo[i].Children) == true {
+			return true
+		}
+	}
+	return false
+}
+
+func (p *ProbeHandler) deleteIfRemoved(currentDevInfo blockdevInfo, newDevInfo []BlockDevice) {
+	blockID, err := currentDevInfo.Node.GetFieldString("BlockID")
+	if err != nil {
+		return
+	}
+	path, err := currentDevInfo.Node.GetFieldString("Path")
+	if err != nil {
+		return
+	}
+	if p.findBlockDev(path, blockID, newDevInfo) == false {
+		p.unregisterBlockdev(path)
+	}
+}
+
+func (p *ProbeHandler) registerBlockdev(blockdev BlockDevice, parentWWN string) *graph.Node {
+	var groupNode *graph.Node
+	childCount := len(blockdev.Children)
+	if childCount > 0 {
+		for i := range blockdev.Children {
+			groupNode = p.registerBlockdev(blockdev.Children[i], blockdev.WWN)
+		}
+		if blockdev.Type == multipathType {
+			groupNode = p.addGroup(blockdev, parentWWN)
```

Comment:
addGroup seems to require the lock on ProbeHandler (line 178). For simplicity, you could do
```
p.Lock()
defer p.Unlock()
```
at the start of the function, like in `unregisterBlockdev`

Reply:
ok